### PR TITLE
Retry window.open if _blank fails

### PR DESCRIPTION
**Pull request by dvoytenko**

### CODE REVIEW

File: ads/alp/handler.js [hunk 1]
```diff
@@ -18,7 +18,7 @@ import {
   addParamToUrl,
   parseQueryString,
 } from '../../src/url';
-import {closest} from '../../src/dom';
+import {closest, openWindowDialog} from '../../src/dom';
```
Comment (cramforce) [inline ads/alp/handler.js]: I'd just call this `windowOpen`, but OK :)

File: src/dom.js [hunk 2]
```diff
@@ -402,3 +403,28 @@ export function hasNextNodeInDocumentOrder(element) {
...
+ * @return {?Window}
+ */
+export function openWindowDialog(win, url, target, opt_features) {
+  // Try first with the specified target.
+  let res = null;
+  try {
+    res = win.open(url, target, opt_features);
+  } catch (e) {
+    dev.error('dom', 'Failed to open url on target: ', target, e);
+    res = null;
```
Comment (cramforce) [inline src/dom.js]: This should not be needed.

File: src/dom.js [hunk 3]
```diff
@@ -402,3 +403,28 @@ export function hasNextNodeInDocumentOrder(element) {
...
+/**
+ * @param {!Window} win
+ * @param {string} url
+ * @param {string} target
+ * @param {string=} opt_features
+ * @return {?Window}
+ */
+export function openWindowDialog(win, url, target, opt_features) {
+  // Try first with the specified target.
+  let res = null;
```
Comment (cramforce) [inline src/dom.js]: Just leave unassigned.

File: src/dom.js [hunk 4]
```diff
@@ -402,3 +403,28 @@ export function hasNextNodeInDocumentOrder(element) {
   } while (currentElement = currentElement.parentNode);
   return false;
 }
+
+
+/**
+ * @param {!Window} win
```
Comment (cramforce) [inline src/dom.js]: Add docs.

File: src/dom.js [hunk 5]
```diff
@@ -402,3 +403,28 @@ export function hasNextNodeInDocumentOrder(element) {
...
+ * @param {string} url
+ * @param {string} target
+ * @param {string=} opt_features
+ * @return {?Window}
+ */
+export function openWindowDialog(win, url, target, opt_features) {
+  // Try first with the specified target.
+  let res = null;
+  try {
+    res = win.open(url, target, opt_features);
```
Comment (cramforce) [inline src/dom.js]: Add lots of references to WKWebView and friends as to why this might fail.

File: src/dom.js [hunk 6]
```diff
@@ -402,3 +403,28 @@ export function hasNextNodeInDocumentOrder(element) {
...
+  } catch (e) {
+    dev.error('dom', 'Failed to open url on target: ', target, e);
+    res = null;
+  }
+
+  // Then try with `_top` target.
+  if (!res && target != '_top') {
+    res = win.open(url, '_top');
```
Comment (cramforce) [inline src/dom.js]: Do you want to retry also when `win.open` did not throw?

File: src/dom.js (hunk 2 above)
Comment (dvoytenko) [inline src/dom.js]: Done

File: src/dom.js (hunk 3 above)
Comment (dvoytenko) [inline src/dom.js]: Done

File: src/dom.js (hunk 4 above)
Comment (dvoytenko) [inline src/dom.js]: Done

File: src/dom.js (hunk 5 above)
Comment (dvoytenko) [inline src/dom.js]: Done

File: src/dom.js (hunk 6 above)
Comment (dvoytenko) [inline src/dom.js]: I do. It retries whenever `win.open` throws or returns `null` or `undefined`.

**Comment by dvoytenko** [comment]
@cramforce All addressed. PTAL.

**Comment by cramforce** [comment]
LGTM

**Pull request merged**